docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,725 @@
+speaking) and users had to explicitly remove unused paths with suicide
+triggers, so any remaining unsatisfied waiting tasks would indicate a problem
+by stalling the workflow.
+
+This document proposes a way to convey the intended path of execution to the
+Cylc 8 spawn-on-demand scheduler by making all task outputs **required** by
+default and giving a new syntax to explicitly mark **optional** outputs. If a
+task finishes without completing all of its required outputs it will be
+retained in the n=0 task pool and flagged as an **incomplete task** (often
+this will coincide with unexpected task failure; otherwise it implies that the
+task did not behave as the workflow configuration expected despite reporting
+success).
+
+*Note: optional triggers were considered and rejected as another way of
+potentially handling this problem - see cylc-admin#128*
+
+## Output Syntax
+
+- `foo:x` means `x` is a *required output* of task `foo`
+  - if a task finishes without completing any required outputs, retain it in
+    the task pool as an *incomplete task*
+
+- `foo:y?` means `y` is an *optional output* of task `foo`
+  - it is OK for a task to finish without completing optional outputs
+
+Abbreviated syntax for the success case:
+- `foo:` mean `foo:succeed` (success is required)
+- `foo?` and `foo:?` mean `foo:succeed?` (success is optional)
+- `foo` means `foo:succeed` UNLESS success is made optional elsewhere in the
+  graph via `foo:succeed?` or `foo:fail?`
+  - this supports Cylc 7 graphs as well as being a sensible default
+  - we should probably give a validation warning here in case the user is
+    actually presuming contradictory success requirements for the same task
+    in different parts of the graph. For complete safety, if success is
+    optional use explicit syntax throughout.
+
+### Interpretation in Trigger Expressions
+
+```
+foo:x => bar
+```
+... means trigger `bar` if `foo` completes output `x`; otherwise
+don't trigger `bar`, and flag `foo` as an incomplete task.
+
+```
+foo:x? => bar
+```
+... means trigger `bar` if `foo` completes output `x`; otherwise
+don't trigger `bar`, but don't flag `foo` as an incomplete task either.
+
+### Caveats
+
+An output can't be both required and optional.
+
+Success should normally be required.
+
+Failure can be optional, in which case success must also be optional, because
+you might need alternate branches to handle:
+- a platform that isn't always available: `foo:submit-fail?`
+- a task that you expect to fail sometimes: `foo:fail?`
+
+Failure can't be required. Expecting a task to fail every time and treating
+it as incomplete if it succeeds would essentially amount to reversing normal 
+success and failure symantics. If you really have an executable that returns
+non-`0` whilst still doing what you need in the workflow, provide a wrapper
+that correctly returns success status.
+- `foo:fail  # ERROR`
+- `foo:submit-fail  # ERROR`
```

Comment:
How are we going to provide Cylc 7 back-compat for this?

As expected `:fail` is commonly used, less expected, I found a fair few workflows using `:submit-fail` in OR conditionals.

Reply:
I think it depends on whether we can infer that the output is optional from the suicide triggers.
I'm not convinced we can do this.
If we can't then I think we should allow this for Cylc 7 workflows (`suite.rc`) and make it an error for Cylc 8 workflows (`flow.cylc`).

Reply:
I'm OK with "failure can't be required" because of the argument that it reverses normal success/fail semantics, which is a variant of the argument that failure is special because by definition it means - or should mean - that the task failed to do what it was supposed to do ... so being "supposed to fail" is perverse.

I think inferring `foo:fail?` in Cylc 7 graphs is quite easy, no need to look at suicide triggers: if both `foo:succeed` and `foo:fail` appear in the graph, they must be optional.  I doubt that anyone is using failure-required semantics (?).

Reply:
(We've gone with `foo:fail` allowed, after discussion on Element)